### PR TITLE
refactor: do script loading more dynamically

### PR DESCRIPTION
**Pull request by fubuloubu**
### What I did
Allow scripts that use click to function correctly. Also load in extra locals correctly using interactive. Basically entire scripting engine overhaul.

fixes: #369

### Checklist

- [x] Passes all linting checks (pre-commit and CI jobs)
- [x] New test cases have been added and are passing
- [x] Documentation has been updated
- [x] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: src/ape_run/_cli.py [hunk 1]
```diff
@@ -1,33 +1,155 @@
...
+
+                ns["main"]()  # Call function
+
+            return call
+
+        else:
+            logger.warning(f"No 'main' method or 'cli' command in script: {relative_filepath}")
+
+            @click.command(short_help=f"Run '{relative_filepath}'")
+            def call():
```
Comment (unparalleled-js) [inline src/ape_run/_cli.py:74]: this call could be a default call above and the `elif` above could call it before doing `ns["main"]()  # Call function`

**Comment by unparalleled-js** [comment]
This seems pretty cool, I am going to try it out

**Comment by unparalleled-js** [comment]
If the script does not exist (like `ape run asdfasdf`), the error you get is:

```
Try 'ape run -h' for help.

Error: No such command 'asdf'.
```

Can we make this somehow more custom, like

```
Try 'ape run <script-name> -h' for help or check if script exists.

Error: No such script or command 'asdf' found.
```

**Comment by unparalleled-js** [comment]
I think my import statements are no longer working in scripts. e.g  a script like:

```python
from .utils import deploy


def main():
    deploy()
```

is failing with

```
  File "/Users/jules/PycharmProjects/ape/src/ape_run/_cli.py", line 62, in call
    eval(code, ns, ns)
  File "/Users/jules/ApeProjects/ape-demo-project/scripts/deploy.py", line 6, in <module>
    from .utils import deploy
KeyError: "'__name__' not in globals"
```

**Comment by fubuloubu** [comment]
> I think my import statements are no longer working in scripts. e.g a script like:
> 
> ```python
> from .utils import deploy
> 
> 
> def main():
>     deploy()
> ```
> 
> is failing with
> 
> ```
>   File "/Users/jules/PycharmProjects/ape/src/ape_run/_cli.py", line 62, in call
>     eval(code, ns, ns)
>   File "/Users/jules/ApeProjects/ape-demo-project/scripts/deploy.py", line 6, in <module>
>     from .utils import deploy
> KeyError: "'__name__' not in globals"
> ```

I just added a small tweak where it ignores "private" files (e.g. `_utils.py`). Give that a try.

Will document this

**Comment by unparalleled-js** [comment]
We could use some documentation around scripts, perhaps in `docs/userguides/projects`, just something like:

```
## Scripts

You can create scripts in the `scripts/` directory that either have a `main()` method or is a simple `click` CLI script.

Here is an example using the `main()` method:
...

Here is an example using the `cli()` method:

```

**Comment by unparalleled-js** [comment]
> I just added a small tweak where it ignores "private" files (e.g. `_utils.py`). Give that a try.
> 
> Will document this

Cool!

I renamed it to `_utils` and tried both `from ._utils import` and `from _utils import` and I am still seeing the same error

**Review by unparalleled-js — DISMISSED** [review]
Thank you for doing this

**Comment by sabotagebeats** [comment]
I'm trying to use this branch but I'm getting this error: 
```
WARNING: Error loading scripts/deploy.py:
        TypeError: Arguments take exactly one parameter declaration, got 2.
INFO: Skipped loading: /home/doge/ape/simple-nft/scripts/deploy.py
```

File: tests/integration/cli/test_run.py [hunk 2]
```diff
@@ -1,36 +1,27 @@
...
 
-@skip_projects(
-    ["empty-config", "no-config", "one-interface", "unregistered-contracts", "test", "geth"]
-)
 def test_run(ape_cli, runner, project):
-    for script_file in project.scripts_folder.glob("*.py"):
-        if script_file.stem == "__init__":
-            continue
+    result = runner.invoke(ape_cli, ["run"])
+    assert result.exit_code == 0, result.output
```
Comment (unparalleled-js) [inline tests/integration/cli/test_run.py:15]: what is the thought behind allowing `ape run` to work without any arguments?

I would think at a script name should be required, no?

File: src/ape_run/_cli.py [hunk 3]
```diff
@@ -1,33 +1,152 @@
...
+    short_help="Run scripts from the `scripts/` folder",
+)
 @click.option(
     "-i",
     "--interactive",
     is_flag=True,
     default=False,
     help="Drop into interactive console session after running",
 )
-@ape_cli_context()
```
Comment (unparalleled-js) [inline src/ape_run/_cli.py:15]: How does the network get set on scripts now?
I usually run ape run <script> --network <network-path>.

Am I going to have to manually add the network_option to all of my scripts that use it now?

File: tests/integration/cli/projects/script/scripts/_utils.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+# Note that any file prepended w/ underscore is ignored
+deploy = lambda: None  # noqa: E731
```
Comment (unparalleled-js) [inline tests/integration/cli/projects/script/scripts/_utils.py:2]:
```suggestion
def deploy():
    pass
```

This is not an anonymous method and there is no reason to use lambda then

File: tests/integration/cli/test_run.py (hunk 2 above)
Comment (unparalleled-js) [inline tests/integration/cli/test_run.py:15]: what is the thought behind allowing ape run to work without any arguments?

I would think a script name should be required, no?

**Review by unparalleled-js — COMMENTED** [review]
I am wondering about something and am about to test it out

File: src/ape_run/_cli.py (hunk 3 above)
Comment (fubuloubu) [reply src/ape_run/_cli.py:15]: if we add it by default, then scripts connect to a network that don't need to

I think focusing more on "a set of functions for building scripts that integrate with ape" should be the goal here

File: tests/integration/cli/test_run.py (hunk 2 above)
Comment (fubuloubu) [reply tests/integration/cli/test_run.py:15]: To display help information about `ape run`, as well as display all the scripts that you can execute from the `scripts/` folder
```bash
$ ape run
  Run scripts from the ``scripts`` folder. ...

Options:
  -i, --interactive  Drop into interactive console session after running
  -h, --help         Show this message and exit.

Commands:
  click   Use a subcommand
  deploy  Run 'scripts/deploy.py:main'
  site    Run 'scripts/site.py'
```

Comment (unparalleled-js) [reply tests/integration/cli/test_run.py:15]: That is neat!

File: src/ape_run/_cli.py [hunk 5]
```diff
@@ -1,33 +1,152 @@
...
 
+        Returns:
+            :class:`~ape.managers.project.ProjectManager`
+        """
 
-@click.command(cls=NetworkBoundCommand, short_help="Run scripts from the `scripts` folder")
-@click.argument("scripts", nargs=-1)
+        from ape import project
+
+        project.config.load()
```
Comment (unparalleled-js) [inline src/ape_run/_cli.py:92]:
```suggestion
        project.config_manager.load()
```

(if this is still needed)

File: tests/integration/cli/projects/script/scripts/_utils.py [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# Note that any file prepended w/ underscore is ignored
```
Comment (unparalleled-js) [inline tests/integration/cli/projects/script/scripts/_utils.py:1]: Can we add a note to the docs about this?

**Review by unparalleled-js — COMMENTED** [review]
The error in the docs happens when I try to run a script,
(but is fixed when renaming `config` to  `config_manager)`
and I am still struggling to import my utils module.

I get:

```
ImportError: attempted relative import with no known parent package
```

I do have a `__init__.py` file so I am not sure why that is.

This is breaking change, so we should get this in before the release!

**Comment by fubuloubu** [comment]
Note: scripts no longer launch networking processes by default, must explicitly connect or use CLI utilities for network bound commands

**Comment by unparalleled-js** [comment]
There are a lot of places throughout the docs that say the `--network` option is available in the `ape run` command, so we need to fix that.

Also still would like docs showing users how to make scripts and differences between the `main()` method approach versus the `cli()` approach.

I have to migrate 90% of my scripts to be CLI based now, since they require the network option. This also means I have to repeat the same code that many times when before I didn't.

**Review by unparalleled-js — APPROVED** [review]
Please fix docs! Otherwise this is nice :)

I will busy for a while migrating 30 scripts 😅 
This is a big breaking change!

**Comment by fubuloubu** [comment]
> There are a lot of places throughout the docs that say the `--network` option is available in the `ape run` command, so we need to fix that.

I think we should ensure this is the approach we'd want to have. It's easy enough for me to add, at least to non-cli backends, by just adding the class that adds the network option to the cli wrapper that the non-cli scripts execute within. Can just as easily add this, and you still have the option of avoiding network usage if you define a cli command without it (since that type will not get injected)

**Comment by unparalleled-js** [comment]
I agree on the approach here for the reasons you have outlined! But we need to fix the places in the docs that still reference the old way, that is all I am saying.

File: docs/userguides/projects.md [hunk 7]
```diff
@@ -49,32 +49,46 @@ a.deploy(project.MyContract)
...
 
 ```bash
-ape run deploy --network ::hardhat
+ape console --network ::hardhat
 ```
 
+## Scripts
+
+You can write scripts that run using the `ape run` command. The `ape run` command will register and run Python
+files defined under the `scripts/` directory, that do not start with an `_` underscore. If the scripts take
```
Comment (unparalleled-js) [inline docs/userguides/projects.md:78]:
```suggestion
files defined under the `scripts/` directory that do not start with an `_` underscore. If the scripts take
```

Comment (unparalleled-js) [inline docs/userguides/projects.md:78]:
```suggestion
files defined under the `scripts/` directory that do not start with an `_` underscore. If the scripts take
```

File: src/ape_run/_cli.py [hunk 8]
```diff
@@ -1,33 +1,155 @@
...
+                except Exception as e:
+                    logger.error_from_exception(
+                        e, f"Exception while parsing script: {relative_filepath}"
+                    )
+                    return None  # Prevents stalling scripts
+
+            self._namespace[filepath.stem] = ns
+            return ns["cli"]
+
+            return None  # NOTE: Allow other scripts to load if loading script breaks
```
Comment (unparalleled-js) [inline src/ape_run/_cli.py:48]: Does this actually ever get called?
Wouldn't the return statement on line 46 always either cause an exception to occur or actually return the CLI?

File: tests/integration/cli/projects/script/scripts/deploy.py [hunk 9]
```diff
@@ -1,6 +1,11 @@
 from ape import networks
 
+# Make sure the re-import bug goes away
+from ._utils import deploy  # noqa: F401
+
 
 def main():
-    assert networks.active_provider.name == "test"
-    print("Script ran!")  # noqa: T001
+    with networks.parse_network_choice("::test"):
```
Comment (unparalleled-js) [inline tests/integration/cli/projects/script/scripts/deploy.py:8]: I don't think this part is needed anymore if `--network` is automatically passed in?
I am not sure how that would affect the test though

**Review by unparalleled-js — APPROVED** [review]
This is a really well-done PR!

I like having the `--network` option on `main`. I know that is a little inconsistent, but the flexibility is still there with the CLI commands and it makes sense to have CLIs be really flexible and main scripts to be classic.

File: src/ape_run/_cli.py (hunk 8 above)
Comment (fubuloubu) [reply src/ape_run/_cli.py:48]: Oh yeah, whoops. I found out that the `if` statement I had before was unreachable, forgot this line!

**Review by unparalleled-js — APPROVED** [review]
LGTM

**Pull request merged**